src/parser/bison_parser.y

```diff
@@ -756,67 +759,137 @@ update_clause : IDENTIFIER '=' expr {
-    $$->setOperations = new std::vector<SetOperation*>();
-  }
-  $$->setOperations->push_back($3);
-  $$->setOperations->back()->nestedSelectStatement = $4;
-  $$->setOperations->back()->resultOrder = $5;
-  $$->setOperations->back()->resultLimit = $6;
-  $$->withDescriptions = $1;
-};
-
-select_within_set_operation : select_with_paren | select_within_set_operation_no_parentheses;
-
-select_within_set_operation_no_parentheses : select_clause { $$ = $1; }
-| select_clause set_operator select_within_set_operation {
-  $$ = $1;
-  if ($$->setOperations == nullptr) {
-    $$->setOperations = new std::vector<SetOperation*>();
-  }
-  $$->setOperations->push_back($2);
-  $$->setOperations->back()->nestedSelectStatement = $3;
-};
-
-select_with_paren : '(' select_no_paren ')' { $$ = $2; }
-| '(' select_with_paren ')' { $$ = $2; };
-
-select_no_paren : select_clause opt_order opt_limit opt_locking_clause {
-  $$ = $1;
-  $$->order = $2;
-
-  // Limit could have been set by TOP.
-  if ($3) {
-    delete $$->limit;
-    $$->limit = $3;
-  }
-
-  if ($4) {
-    $$->lockings = $4;
-  }
-}
-| select_clause set_operator select_within_set_operation opt_order opt_limit opt_locking_clause {
-  $$ = $1;
-  if ($$->setOperations == nullptr) {
-    $$->setOperations = new std::vector<SetOperation*>();
-  }
-  $$->setOperations->push_back($2);
-  $$->setOperations->back()->nestedSelectStatement = $3;
-  $$->setOperations->back()->resultOrder = $4;
-  $$->setOperations->back()->resultLimit = $5;
-  $$->lockings = $6;
-};
+select_statement:
+          query_expression
+        | query_expression_parens
+        ;
+
+query_expression:
+          query_expression_body opt_order opt_limit opt_locking_clause
+          {
+            if ($1->setOperations == nullptr) {
+              $1->order = $2;
+
+              // Limit could have been set by TOP.
+              if ($3 != nullptr) {
+                delete $1->limit;
+                $1->limit = $3;
+              }
+
+              if ($4 != nullptr) {
+                $1->lockings = $4;
+              }
+            }
+            else {
+              $1->setOperations->back()->resultOrder = $2;
+              $1->setOperations->back()->resultLimit = $3;
+            }
+
+            $$ = $1;
+          }
+        | with_clause query_expression_body opt_order opt_limit opt_locking_clause
+          {
+            $2->withDescriptions = $1;
+            if ($2->setOperations == nullptr) {
+              $2->order = $3;
+
+              // Limit could have been set by TOP.
+              if ($4 != nullptr) {
+                delete $2->limit;
+                $2->limit = $4;
+              }
+
+              if ($5 != nullptr) {
+                $2->lockings = $5;
+              }
+            }
+            else {
+              $2->setOperations->back()->resultOrder = $3;
+              $2->setOperations->back()->resultLimit = $4;
+            }
+
+            $$ = $2;
+          }
+
+/*
```

Comment:
knit-picky: We either use C-style comments with an asterisk `*` on each line or C++-style single line comments.

Reply:
*nit-pick ;)